src/lru.rs

```diff
@@ -0,0 +1,343 @@
+    fn lru_index(&self) -> &LruIndex;
+}
+
+#[derive(Debug)]
+pub(crate) struct LruIndex {
+    /// Index in the approprate LRU list, or std::usize::MAX if not a
+    /// member.
+    index: AtomicUsize,
+}
+
+impl<Node> Default for Lru<Node>
+where
+    Node: LruNode,
+{
+    fn default() -> Self {
+        Lru::new()
+    }
+}
+
+impl<Node> Lru<Node>
+where
+    Node: LruNode,
+{
+    /// Creates a new LRU list where LRU caching is disabled.
+    pub fn new() -> Self {
+        Lru {
+            green_zone: AtomicUsize::new(0),
+            data: Mutex::new(LruData::new()),
+        }
+    }
+
+    #[cfg_attr(not(test), allow(dead_code))]
+    fn with_seed(seed: &str) -> Self {
+        Lru {
+            green_zone: AtomicUsize::new(0),
+            data: Mutex::new(LruData::with_seed(seed)),
+        }
+    }
+
+    /// Adjust the total number of nodes permitted to have a value at
+    /// once.  If `len` is zero, this disables LRU caching completely.
+    pub fn set_lru_capacity(&self, len: usize) {
+        let mut data = self.data.lock();
+
+        // We require each zone to have at least 1 slot. Therefore,
+        // the length cannot be just 1 or 2.
+        if len == 0 {
+            self.green_zone.store(0, Ordering::Release);
+            data.resize(0, 0, 0);
+        } else {
+            let len = std::cmp::max(len, 3);
+
+            // Top 10% is the green zone. This must be at least length 1.
+            let green_zone = std::cmp::max(len / 10, 1);
+
+            // Next 20% is the yellow zone.
+            let yellow_zone = std::cmp::max(len / 5, 1);
+
+            // Remaining 70% is the red zone.
+            let red_zone = len - yellow_zone - green_zone;
+
+            // We need quick access to the green zone.
+            self.green_zone.store(green_zone, Ordering::Release);
+
+            // Resize existing array.
+            data.resize(green_zone, yellow_zone, red_zone);
+        }
+    }
+
+    /// Records that `node` was used. This may displace an old node (if the LRU limits are
+    pub fn record_use(&self, node: &Arc<Node>) -> Option<Arc<Node>> {
+        log::debug!("record_use(node={:?})", node);
+
+        // Load green zone length and check if the LRU cache is even enabled.
+        let green_zone = self.green_zone.load(Ordering::Acquire);
+        log::debug!("record_use: green_zone={}", green_zone);
+        if green_zone == 0 {
+            return None;
+        }
+
+        // Find current index of list (if any) and the current length
+        // of our green zone.
+        let index = node.lru_index().load();
+        log::debug!("record_use: index={}", index);
+
+        // Already a member of the list, and in the green zone -- nothing to do!
+        if index < green_zone {
+            return None;
+        }
+
+        self.data.lock().record_use(node)
+    }
+}
+
+impl<Node> LruData<Node>
+where
+    Node: LruNode,
+{
+    fn new() -> Self {
+        Self::with_rng(SmallRng::from_entropy())
```

Comment:
Let's just use a fixed seed here: reproducibility is nice!

Also (and this is a bit of rant), I feel like `rand` is a bit of an overkill. It's a big crate with many dependencies. I often feel that just hard-coding xor-shift, as stdlib does, would be better:

https://github.com/rust-lang/rust/blob/dbec74ffa7982de8e066a93969ce70a891d7908b/src/libcore/slice/sort.rs#L486-L493

Reply:
Switched to fixed seed. I agree rand is probably overkill but I'm not inclined to "roll our own" -- though I'd probably merge a PR if you feel that strongly. =) 

(A "max-min-rand" might be a nice crate to publish)